### PR TITLE
Feature/new configuration refactor

### PR DESCRIPTION
**Pull request by ficklephil**
New Configuration using Node Config.
Old configuration removed.
Unit tests fixed up.
New values added to values.yaml and values.preview.template.yaml
How to run locally documented in Readme.md.

### CODE REVIEW

File: api/application.ts [hunk 1]
```diff
@@ -34,12 +36,14 @@ app.use(
         cookie: {
             httpOnly: true,
             maxAge: 1800000,
-            secure: config.secureCookie !== false,
+            // TODO: Possible to deprecate?
+            // secure: config.secureCookie !== false,
+            secure: false,
```
Comment (paddybasi) [inline api/application.ts]: We shouldn't deprecate this, in fact if anything we need to enable by default and just disable in lower environments (e.g. preview etc)

File: api/configuration/index.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ */
+export const getConfigValue = reference => {
+  return config.get<string>(reference)
+}
+
+export const hasConfigValue = reference => {
+  return config.has(reference)
+}
+
+export const getConfigValueNumber = reference => {
```
Comment (paddybasi) [inline api/configuration/index.ts]: Let's perhaps have a templated function instead:

```typescript
export const getConfigValue<T> = reference => {
   return config.get<T>(reference)
}
```

File: api/configuration/index.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  return config.get<number>(reference)
+}
+
+export const getIDamSecret = (secretsConfig): string => {
```
Comment (paddybasi) [inline api/configuration/index.ts]: we no longer have a secretsConfig so possibly remove this?

File: api/lib/log4jui.ts [hunk 4]
```diff
@@ -20,7 +37,8 @@ export function leftPad(str: string, length = 20): string {
 
 export function getLogger(category: string): JUILogger {
     const logger: log4js.Logger = log4js.getLogger(category)
-    logger.level = config.logging || 'off'
+    // @ts-ignore
```
Comment (paddybasi) [inline api/lib/log4jui.ts]: why is this needed?

File: api/local.ts [hunk 5]
```diff
@@ -1,6 +1,60 @@
...
+console.log(getConfigValue(MAX_LOG_LINE))
+console.log(getConfigValue(LOGGING))
+console.log(getConfigValue(APP_INSIGHTS_KEY))
+console.log(getConfigValue(SERVICES_TERMS_AND_CONDITIONS_URL))
+console.log(getConfigValue(HEALTH))
+console.log('END CHECK OF ENVIRONMENTAL VARIABLES')
+
+console.log(getConfigValue(S2S_SECRET))
+console.log(getConfigValue(APP_INSIGHTS_SECRET))
+const mountedSecrets = propertiesVolume.addTo(config)
```
Comment (paddybasi) [inline api/local.ts]: this essentially does the same now as initialiseSecrets()

File: api/services/idam.ts [hunk 6]
```diff
@@ -1,14 +1,25 @@
...
 import { valueOrNull } from '../lib/util'
 
-const url = config.services.idam.idamApiUrl
-
-const idamSecret = process.env.IDAM_SECRET || 'AAAAAAAAAAAAAAAA'
-const idamClient = config.idamClient
-const idamProtocol = config.protocol
-const oauthCallbackUrl = config.services.idam.oauthCallbackUrl
+const url = getConfigValue(SERVICES_IDAM_API_URL)
+const mountedSecrets = propertiesVolume.addTo(config)
```
Comment (paddybasi) [inline api/services/idam.ts]: again do we need this here?

File: api/services/serviceAuth.ts [hunk 7]
```diff
@@ -1,24 +1,38 @@
...
 import { http } from '../lib/http'
 import * as log4jui from '../lib/log4jui'
 
-const url = config.services.s2s
-const microservice = config.microservice
-const s2sSecret = process.env.S2S_SECRET || 'AAAAAAAAAAAAAAAA'
+const s2sPath = getConfigValue(SERVICE_S2S_PATH)
+const microservice = getConfigValue(MICROSERVICE)
+
+initialiseSecrets()
```
Comment (paddybasi) [inline api/services/serviceAuth.ts]: do we need this here?

File: charts/xui-webapp/values.preview.template.yaml [hunk 8]
```diff
@@ -4,6 +4,30 @@ nodejs:
   environment:
        IDAM_URL: https://idam-api.aat.platform.hmcts.net
        S2S_URL: http://rpe-service-auth-provider-aat.service.core-compute-aat.internal
+       APPINSIGHTS_INSTRUMENTATIONKEY: ${APPINSIGHTS_INSTRUMENTATIONKEY}
```
Comment (paddybasi) [inline charts/xui-webapp/values.preview.template.yaml]: this should reside in the keyVault (and pulled through as an env variable)

File: charts/xui-webapp/values.preview.template.yaml [hunk 9]
```diff
@@ -4,6 +4,30 @@ nodejs:
...
        IDAM_SECRET: ${IDAM_SECRET}
        NODE_TLS_REJECT_UNAUTHORIZED: '0'
+       # Health Endpoints
+       HEALTH_CCD_COMPONENT_API: http://ccd-api-gateway-web-aat.service.core-compute-aat.internal/health
+       HEALTH_CCD_DATA_API: http://ccd-data-store-api-aat.service.core-compute-aat.internal/health
+       HEALTH_DOCUMENTS_API: http://dm-store-aat.service.core-compute-aat.internal/health
+       HEALTH_EM_ANNO_API: http://em-anno-aat.service.core-compute-aat.internal/health
+       HEALTH_TERMS_AND_CONDITIONS_API: http://xui-terms-and-conditions-aat.service.core-compute-aat.internal/health
+       # Services CCD Endpoints
+       SERVICES_CCD_COMPONENT_API: http://ccd-api-gateway-web-aat.service.core-compute-aat.internal
```
Comment (paddybasi) [inline charts/xui-webapp/values.preview.template.yaml]: ideally we need to make these as 'templated' variables

File: config/custom-environment-variables.json [hunk 10]
```diff
@@ -0,0 +1,50 @@
+{
+  "environment": {
+    "name": "This file: custom-environment-variables overwrites all other .json files within the /config folder"
+  },
+  "cookies": {
+    "token": "__auth__",
+    "userId": "__userid__",
+    "sessionId": "__sessionId__"
+  },
+  "appInsights": {
```
Comment (paddybasi) [inline config/custom-environment-variables.json]: have this as .e.g
```config
'secrets.rpx.appInsightsInstrumentationKey'
```

File: package.json [hunk 11]
```diff
@@ -61,24 +61,30 @@
...
     "applicationinsights": "^1.5.0",
     "applicationinsights-js": "^1.0.20",
     "axios": "^0.18.0",
     "class-transformer": "^0.1.10",
     "classlist.js": "^1.1.20150312",
+    "config": "^3.2.5",
     "cookie-parser": "^1.4.4",
     "core-js": "^2.5.4",
     "crypto-js": "^3.1.9-1",
+    "dotenv-extended": "^2.7.1",
```
Comment (paddybasi) [inline package.json]: if we're using this package, we can make it a bit more dev friendly by having the .env.defaults

File: api/local.ts (hunk 5 above)
Comment (udaydenduluri33) [reply api/local.ts]: This will be done in my next PR

File: api/services/serviceAuth.ts (hunk 7 above)
Comment (udaydenduluri33) [reply api/services/serviceAuth.ts]: This will be resolved in my next PR

File: charts/xui-webapp/values.preview.template.yaml (hunk 9 above)
Comment (udaydenduluri33) [reply charts/xui-webapp/values.preview.template.yaml]: This is a clean-up task. But for a later day

File: charts/xui-webapp/values.preview.template.yaml (hunk 8 above)
Comment (udaydenduluri33) [reply charts/xui-webapp/values.preview.template.yaml]: Please wait for the next PR.

File: package.json (hunk 11 above)
Comment (udaydenduluri33) [reply package.json]: Nice to have sir. Lets do it in re-factor exercise

File: api/application.ts (hunk 1 above)
Comment (udaydenduluri33) [reply api/application.ts]: This a done now. please have a look @paddybasi

File: api/configuration/index.ts (hunk 2 above)
Comment (udaydenduluri33) [reply api/configuration/index.ts]: Done @paddybasi

File: api/lib/log4jui.ts (hunk 4 above)
Comment (udaydenduluri33) [reply api/lib/log4jui.ts]: Done @paddybasi

File: api/application.ts [hunk 12]
```diff
@@ -20,26 +25,23 @@ import routes from './routes'
...
 
 app.set('trust proxy', 1)
-
 app.use(
     session({
         cookie: {
             httpOnly: true,
             maxAge: 1800000,
-            secure: config.secureCookie !== false,
+            secure: getConfigValue(SECURE_COOKIE) !== false,
```
Comment (paddybasi) [inline api/application.ts]: having the
```javascript
!== false
```
doesn't make much sense anymore, you can remove it if you wish (the values are now json objects, so true booleans or at least they should be)

File: api/application.ts [hunk 13]
```diff
@@ -53,35 +55,8 @@ app.use(errorStack)
 app.use(bodyParser.json())
 app.use(bodyParser.urlencoded({extended: true}))
 
-// TODO: remove this when we have proper frontend configuration
```
Comment (paddybasi) [inline api/application.ts]: is this cookie no longer needed? just curious

File: api/auth/manageCasesUserRoleAuth.ts [hunk 14]
```diff
@@ -9,7 +7,7 @@ import {config} from '../config'
  *
  * @returns {string} - ie. 'caseworker'
  */
-export const getLoginRoleMatcher = () => config.loginRoleMatcher
+export const getLoginRoleMatcher = () => 'caseworker'
```
Comment (paddybasi) [inline api/auth/manageCasesUserRoleAuth.ts]: should this not also be in config?

File: api/lib/appInsights.ts [hunk 15]
```diff
@@ -9,7 +12,8 @@ const environment = process.env.XUI_ENV || 'local'
 
 if (environment !== 'local') {
     applicationinsights
-        .setup(config.appInsightsInstrumentationKey)
+        // @ts-ignore
```
Comment (paddybasi) [inline api/lib/appInsights.ts]: is this needed?

File: api/lib/log4jui.ts [hunk 16]
```diff
@@ -8,10 +12,23 @@ import { isReqResSet, request } from './middleware/responseRequest'
 
 // the longest category length we have currently
 const maxCatLength = 14
-const sessionid = config.cookies.sessionId
+const sessionid = getConfigValue(COOKIES_SESSION_ID)
 
 // This is done to mimic log4js calls
-log4js.configure(config.log4jui)
+log4js.configure({
```
Comment (paddybasi) [inline api/lib/log4jui.ts]: can we move this out to default.json

File: api/application.ts (hunk 13 above)
Comment (udaydenduluri33) [reply api/application.ts]: Discussed @ficklephil  and he said its not being used anywhere

File: api/application.ts (hunk 12 above)
Comment (udaydenduluri33) [reply api/application.ts]: Done

File: api/lib/appInsights.ts (hunk 15 above)
Comment (udaydenduluri33) [reply api/lib/appInsights.ts]: No, removed it. @paddybasi

File: api/auth/manageCasesUserRoleAuth.ts (hunk 14 above)
Comment (udaydenduluri33) [reply api/auth/manageCasesUserRoleAuth.ts]: Yes, agreed. Done. @paddybasi

File: api/lib/log4jui.ts (hunk 16 above)
Comment (udaydenduluri33) [reply api/lib/log4jui.ts]: Done. @paddybasi
